Allow access to semantic model during syntax walk

chsienki opened the pull request:
This is an implementation for https://github.com/dotnet/roslyn/issues/49873 

Essentially we allow for a new kind of walker than takes a context, instead of just the node itself. On that context we pass the node and a semantic model for the tree the node belongs to. In this way we can re-use the same semantic model for all walkers of the tree, without keeping it alive any longer than we need to.

User can query the model at the point of node acquisition, allowing them to know upfront whether to keep the node or discard it, and no longer need to access the semantic model in the execute method itself.

It's not a particularly complicated PR, but I've broken it out into a couple of commits that should help explain the approach.

chsienki commented on src/Compilers/Core/Portable/SourceGeneration/ISyntaxReceiver.cs:
```diff
@@ -5,7 +5,18 @@
+    /// <remarks>
+    /// This interface is used internally in Roslyn to support backwards compatibility, and should not be implemented directly by consumers.
+    /// </remarks>
+    public interface ISyntaxReceiverBase
```
This should make the implementation strategy backwards compatible.

chsienki commented on src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:
```diff
@@ -61,6 +62,11 @@ internal GeneratorExecutionContext(Compilation compilation, ParseOptions parseOp
+        /// <summary>
+        /// If the generator registered an <see cref="ISyntaxReceiverWithContext"/> during initialization, this will be the instance created for this generation pass.
+        /// </summary>
+        public ISyntaxReceiverWithContext? SyntaxReceiverWithContext { get; }
```
We can't have a single field without breaking back compat. An alternative approach could be to add a method to the context such as `GetReceiver<T>() where T : ISyntaxReceiverBase`

jaredpar replied:
Couldn't we have a single field and just implement each property `get` as the cast operation?

Not sure it's worth the trade off though.

chsienki replied:
Yeah, we'd still need two public properties though, so its ergonomically not really any different for the end user.

---
In reply to: [562023044](https://github.com/dotnet/roslyn/pull/50419#discussion_r562023044) [](ancestors = 562023044)

chsienki commented:
Note: I need to add tests, but want to verify the approach first.

CyrusNajmabadi commented on src/Compilers/Core/Portable/SourceGeneration/ISyntaxReceiver.cs:
```diff
@@ -22,8 +33,11 @@ namespace Microsoft.CodeAnalysis
+    /// <see cref="ISyntaxReceiverWithContext" />, not both.
     /// </remarks>
-    public interface ISyntaxReceiver
+    public interface ISyntaxReceiver : ISyntaxReceiverBase
```
can we obsolete this and push people to use ISyntaxReceiverWithContext ?

chsienki replied:
Yep, we can definitely take that route.

333fred replied:
I would also push for obsoleting this API, but we do need compat council approval. Please send an email.

chsienki commented:
@dotnet/roslyn-compiler for review please.

333fred commented on src/Compilers/Core/Portable/SourceGeneration/GeneratorSyntaxWalker.cs:
```diff
@@ -2,20 +2,41 @@
+        public void VisitWithModel(SemanticModel model, SyntaxNode node)
+        {
+            Debug.Assert(_semanticModel is null);
+            _semanticModel = model;
```
Consider asserting that, if `_syntaxReceiver` is a `ISyntaxReceiverWithContext`, `model` is not null. Or just that `model` is not null at all.

333fred commented on src/Compilers/Core/Portable/SourceGeneration/ISyntaxReceiver.cs:
```diff
@@ -36,5 +50,35 @@ public interface ISyntaxReceiver
     /// </summary>
     /// <returns>An instance of an <see cref="ISyntaxReceiver"/></returns>
-    public delegate ISyntaxReceiver SyntaxReceiverCreator();
+    public delegate ISyntaxReceiverBase SyntaxReceiverCreator();
```
Is this a breaking change?

chsienki replied:
Technically it's a breaking public API change. Anything that uses a `SyntaxReceiverCreator` instance directly to create a receiver would potentially break, as the return type has changed.

In reality we would expect that only the generator driver internals are doing that. If the calling code is just passing an instance of the creator it's not a breaking change to change the return type. It can still return a more derived type and be a valid delegate.

IMO, its fine. Obviously if compat disagrees, we can add a new overload with a new delegate to maintain strict back compat.

jaredpar replied:
I would prefer we went with a new `delegate` here to avoid the potential for corner case compat breaks.

333fred commented:
Done review pass (commit 3)

jaredpar commented on src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:
```diff
@@ -148,6 +154,28 @@ private static void CheckIsEmpty<T>(T x)
+    /// <summary>
+    /// Context passed to an <see cref="ISyntaxReceiverWithContext"/> when <see cref="ISyntaxReceiverWithContext.OnVisitSyntaxNode(GeneratorSyntaxReceiverContext)"/> is called
+    /// </summary>
+    public struct GeneratorSyntaxReceiverContext
```
```suggestion
    public readonly struct GeneratorSyntaxReceiverContext
```

jaredpar commented on src/Compilers/Core/Portable/SourceGeneration/GeneratorSyntaxWalker.cs:
```diff
@@ -2,20 +2,41 @@
         }
 
+        public void VisitWithModel(SemanticModel model, SyntaxNode node)
+        {
```
Should we assert here that `model.SyntaxTree == node.SyntaxTree`?

chsienki commented:
@jaredpar @333fred based on the feedback here I've split the creator function into two delegates. Given that the 'base' interface would only be used internally, I've removed it and now just using `object` instead.

I also renamed things to be a little less clunky: an `ISyntaxReceiver` gets a `SyntaxNode` and now `ISyntaxContextReceiver` gets a `GeneratorSyntaxContext`.

333fred commented:
@chsienki looks like there's build failures.

chsienki commented:
> @chsienki looks like there's build failures.

Yeah, its was a failure in the tests. Fixed, and added some tests, still working on adding more.

jaredpar commented on src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:
```diff
@@ -135,17 +141,70 @@ internal void RegisterForAdditionalFileChanges(EditCallback<AdditionalFileEdit>
+            if (builder.SyntaxReceiverCreator is object || builder.SyntaxReceiverWithContextCreator is object)
+            {
+                // only a single SyntaxReceiverCreator / SyntaxContextReceiverCreator can be registered per generator
+                throw new InvalidOperationException(string.Format(CodeAnalysisResources.Single_type_per_generator_0, $"{nameof(SyntaxReceiverCreator)} / {nameof(SyntaxContextReceiverCreator)}"));
```
Why do we have this restriction? I can't exactly see a good reason to have both but it doesn't seem inherently wrong to do so either.

chsienki replied:
We didn't specifically need it before, but now that we're using an adaptor you can't supply both.

jaredpar commented on src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs:
```diff
@@ -178,21 +178,28 @@ internal GeneratorDriverState RunGeneratorsCore(Compilation compilation, Diagnos
-                // create the syntax receiver if requested
-                if (generatorState.Info.SyntaxReceiverCreator is object)
+                // create the syntax receivers if needed
+                Debug.Assert(!(generatorState.Info.SyntaxReceiverCreator is object && generatorState.Info.SyntaxContextReceiverCreator is object));
```
It seems awkward that throughout this change we have to constantly be validating that we have only one of these set. It adds this complexity throughout the code base. 

Did you consider instead the idea of just making `ISyntaxReceiver` an implementation detail of the registration process? Consider that you could define an adapter class that implements `ISyntaxReceiverWithContext` and just wraps a `ISyntaxReceiver` instance. The registration layer could apply this adapter and the rest of the code base could essentially be ignorant of the fact that we have `ISyntaxReceiver` at all.

chsienki replied:
Yeah I think that might be a better way to go, we still need to 'unwrap' the adaptor if you request it in the generator context, but that seems simple enough. Let me play with that idea a bit.

jaredpar commented on src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:
```diff
@@ -157,30 +157,21 @@ public void RegisterForSyntaxNotifications(SyntaxReceiverCreator receiverCreator
         }
 
-        private static void CheckIsEmpty<T>(T x)
+        private static void CheckIsEmpty<T>(T x, string? typeName = null)
```
```suggestion
        private static void CheckIsEmpty<T>(T x, string? typeName = null) where T : class
```

jaredpar commented on src/Compilers/Core/Portable/SourceGeneration/SyntaxContextReceiverAdaptor.cs:
```diff
@@ -0,0 +1,23 @@
+    /// <summary>
+    /// Wraps an <see cref="ISyntaxReceiver"/> in an <see cref="ISyntaxContextReceiver"/>
+    /// </summary>
+    internal class SyntaxContextReceiverAdaptor : ISyntaxContextReceiver
```
```suggestion
    internal sealed class SyntaxContextReceiverAdaptor : ISyntaxContextReceiver
```

chsienki commented:
@jaredpar Thanks, the adaptor approach ends up much cleaner. I also noted an interesting issue with how we were handling `null` return values from the creator func (we weren't so it was just failing down the line). I've updated it to just not register in those cases, but I'm open to suggestions on the behaviour here.

chsienki commented on src/Compilers/CSharp/Test/Semantic/SourceGeneration/SyntaxAwareGeneratorTests.cs:
```diff
@@ -417,21 +567,83 @@ void Function()
 
-        class TestSyntaxReceiver : ISyntaxReceiver
+        [Fact]
+        public void Syntax_Receiver_Return_Null_During_Creation()
```
Added tests to cover the 'return null' behavior of the creator funcs.

chsienki commented on src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:
```diff
@@ -19,13 +19,14 @@ namespace Microsoft.CodeAnalysis
             AnalyzerConfigOptions = optionsProvider;
-            SyntaxReceiver = syntaxReceiver;
+            SyntaxReceiver = (syntaxReceiver as SyntaxContextReceiverAdaptor)?.Receiver;
+            SyntaxContextReceiver = (syntaxReceiver is SyntaxContextReceiverAdaptor) ? null : syntaxReceiver;
```
I'm explicitly nulling out the ContextReceiver when it's an adaptor so we don't expose that detail to users directly.

cston commented on src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:
```diff
@@ -135,19 +141,63 @@ internal void RegisterForAdditionalFileChanges(EditCallback<AdditionalFileEdit>
+        /// </summary>
+        /// <remarks>
+        /// This method allows generators to be 'syntax aware'. Before each generation the <paramref name="receiverCreator"/> will be invoked to create
+        /// an instance of <see cref="ISyntaxReceiver"/>. This receiver will have its <see cref="ISyntaxContextReceiver.OnVisitSyntaxNode(GeneratorSyntaxContext)"/> 
```
Should this be "an instance of ISyntaxContextReceiver"?

cston commented on src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:
```diff
@@ -135,19 +141,63 @@ internal void RegisterForAdditionalFileChanges(EditCallback<AdditionalFileEdit>
+        /// A new instance of <see cref="ISyntaxContextReceiver"/> is created prior to every call to <see cref="ISourceGenerator.Execute(GeneratorExecutionContext)"/>, 
+        /// meaning there is no need to manage the lifetime of the receiver or its contents.
+        /// </remarks>
+        /// <param name="receiverCreator">A <see cref="SyntaxContextReceiverCreator"/> that can be invoked to create an instance of <see cref="ISyntaxReceiver"/></param>
```
"an instance of ISyntaxContextReceiver"?

cston commented on src/Compilers/Core/Portable/SourceGeneration/SyntaxContextReceiverAdaptor.cs:
```diff
@@ -0,0 +1,23 @@
+    /// </summary>
+    internal sealed class SyntaxContextReceiverAdaptor : ISyntaxContextReceiver
+    {
+        public SyntaxContextReceiverAdaptor(ISyntaxReceiver? receiver)
```
Why allow a `null` `receiver`?

jaredpar replied:
As written this is accounting for the case where the user delegate returns `null`. Could move that logic up into the code which creates the adapter and avoid the `null` here.

jaredpar commented on src/Compilers/CSharp/Test/Semantic/SourceGeneration/SyntaxAwareGeneratorTests.cs:
```diff
@@ -83,22 +84,130 @@ class C { }
+
+            Assert.Single(compilation.SyntaxTrees);
+
+            ISyntaxContextReceiver? receiver = null;
```
Consider having two locals here: `ISyntaxContextReceiver` and `ISyntaxReceiver` then read them both in the lambda callback. That way the `Assert` at the end can be a pair of `Null` and `NotNull`. The presence of `NotNull` ensures the lambda is executing as expected where as the current form of the test could succeed if the lambda doesn't execute.

cston commented on src/Compilers/Core/Portable/SourceGeneration/SyntaxContextReceiverAdaptor.cs:
```diff
@@ -11,13 +11,24 @@ namespace Microsoft.CodeAnalysis
+                return new SyntaxContextReceiverAdaptor(rx);
+            }
+            // in the case that the creator function returns null, we'll also return a null adaptor
+            return null!;
```
It seems surprising to return `null!`. Consider checking the `creator` in the caller of this method and only calling when not `null`.

cston replied:
Consider changing the method return type to `SyntaxContextReceiverCreator?`.

cston commented on src/Compilers/Core/Portable/SourceGeneration/SyntaxContextReceiverAdaptor.cs:
```diff
@@ -0,0 +1,34 @@
+            }
+            // in the case that the creator function returns null, we'll also return a null adaptor
+            return null!;
+        };
```
Trailing `;`

cston commented on src/Compilers/Core/Portable/SourceGeneration/ISyntaxReceiver.cs:
```diff
@@ -37,4 +40,41 @@ public interface ISyntaxReceiver
+    /// Allows a generator to provide instances of an <see cref="ISyntaxContextReceiver"/>
+    /// </summary>
+    /// <returns>An instance of an <see cref="ISyntaxContextReceiver"/></returns>
+    public delegate ISyntaxContextReceiver SyntaxContextReceiverCreator();
```
Should the delegate return type be `ISyntaxContextReceiver?` instead since we handle `null` return values?

chsienki replied:
Yeah, this was one of the questions I had about the null handling. We can explicitly allow `null`, ignore `null` values, or throw when provided with a `null`. The disadvantage of throwing is it won't occur until much later in the process, so seems disconnected. I think you're probably right that we should just explicitly allow `null` values via annotation. 

---
In reply to: [564862256](https://github.com/dotnet/roslyn/pull/50419#discussion_r564862256) [](ancestors = 564862256)

jaredpar commented:
@chsienki did this merge make the snap or do we need to manually port to release/dev16.9?
